pkg/scheduler/actions/allocate/allocate.go

```diff
@@ -228,101 +385,123 @@ func (alloc *Action) allocateResourcesForTasks(tasks *util.PriorityQueue, job *a
-				task.Namespace, task.Name, bestNode.Name)
-
-			// Allocate releasing resource to the task if any.
-			if task.InitResreq.LessEqual(bestNode.FutureIdle(), api.Zero) {
-				klog.V(3).Infof("Pipelining Task <%v/%v> to node <%v> for <%v> on <%v>",
-					task.Namespace, task.Name, bestNode.Name, task.InitResreq, bestNode.Releasing)
-				if err := stmt.Pipeline(task, bestNode.Name, false); err != nil {
-					klog.Errorf("Failed to pipeline Task %v on %v in Session %v for %v.",
-						task.UID, bestNode.Name, ssn.UID, err)
-					if rollbackErr := stmt.UnPipeline(task); rollbackErr != nil {
-						klog.Errorf("Failed to unpipeline Task %v on %v in Session %v for %v.",
-							task.UID, bestNode.Name, ssn.UID, rollbackErr)
-					}
-				} else {
-					metrics.UpdateE2eSchedulingDurationByJob(job.Name, string(job.Queue), job.Namespace, metrics.Duration(job.CreationTimestamp.Time))
-					metrics.UpdateE2eSchedulingLastTimeByJob(job.Name, string(job.Queue), job.Namespace, time.Now())
-				}
+			klog.V(5).Infof("Predicate filtered node %v, idle: %v and future idle: %v do not meet the requirements of task: %v",
+				n.Name, n.Idle, n.FutureIdle(), task.Name)
+		}
+	}
+	candidateNodes = append(candidateNodes, idleCandidateNodes)
+	candidateNodes = append(candidateNodes, futureIdleCandidateNodes)
+
+	var bestNode *api.NodeInfo
+	var higestScore float64
+	for index, nodes := range candidateNodes {
+		if klog.V(5).Enabled() {
+			for _, node := range nodes {
+				klog.V(5).Infof("node %v, idle: %v, future idle: %v", node.Name, node.Idle, node.FutureIdle())
+			}
+		}
+		switch {
+		case len(nodes) == 0:
+			klog.V(5).Infof("Task: %v, no matching node is found in the candidateNodes（index: %d） list.", task.Name, index)
+		case len(nodes) == 1: // If only one node after predicate, just use it.
+			bestNode = nodes[0]
+		case len(nodes) > 1: // If more than one node after predicate, using "the best" one
+			nodeScores := util.PrioritizeNodes(task, nodes, ssn.BatchNodeOrderFn, ssn.NodeOrderMapFn, ssn.NodeOrderReduceFn)
+
+			bestNode = ssn.BestNodeFn(task, nodeScores)
+			if bestNode == nil {
+				bestNode, higestScore = util.SelectBestNodeAndScore(nodeScores)
 			}
 		}
 
-		if ssn.JobReady(job) && !tasks.Empty() {
-			jobs.Push(job)
+		// If a proper node is found in idleCandidateNodes, skip futureIdleCandidateNodes and directly return the node information.
+		if bestNode != nil {
 			break
 		}
 	}
+	return bestNode, higestScore
+}
 
-	if ssn.JobReady(job) {
-		stmt.Commit()
-	} else {
-		if !ssn.JobPipelined(job) {
-			stmt.Discard()
+func (alloc *Action) allocateResourcesForTask(stmt *framework.Statement, task *api.TaskInfo, node *api.NodeInfo, job *api.JobInfo) {
+	// Allocate idle resource to the task.
+	if task.InitResreq.LessEqual(node.Idle, api.Zero) {
+		klog.V(3).Infof("Binding Task <%v/%v> to node <%v>", task.Namespace, task.Name, node.Name)
+		if err := stmt.Allocate(task, node); err != nil {
```

Comment:
Missed UnAllocate? The master branch now has merged UnAllocate here:
https://github.com/volcano-sh/volcano/blob/dca41600e51e5aa667396eed28c8a52db718b981/pkg/scheduler/actions/allocate/allocate.go#L282-L288

Reply:
This is a seperate branch, and will rebase master when merged into master.